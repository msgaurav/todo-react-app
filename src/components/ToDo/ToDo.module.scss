.ToDo_wrapper {
    max-width: 500px;
    margin: auto;
    padding: 3rem 1rem 7rem;

    .form_container {
        position: relative;
        // For linear-gradient border ...
        background-image: linear-gradient(350.69deg ,#3beecd 13.42%,#ff4ecd 42.37%,#1a75ff 103.09%);
        border-radius: 3px;
        padding: 2px;

        .inputForm_element {
            border: none;
            font-size: 1.1rem;
            width: 100%;
            padding: 10px 46px 10px 10px;

            &:focus, &:hover {
                outline: none;
            }
        }
        
        .clearInput_icon {
            position: absolute;
            right: 10px;
            top: 10px;
            cursor: pointer;
            color: red;
        }
    }

    .taskList_container {
        margin: 2.5rem 0;

        .tools_container {
            text-align: right;

            .colorPicker {
                height: 36px;
                padding: 1px 3px;
                vertical-align: bottom;
                cursor: pointer;
                box-shadow: inset 2px 2px 3px rgba(255, 255, 255, .6),
                inset -2px -2px 3px rgba(0, 0, 0, .6);

                &:active {
                    box-shadow: inset -2px -2px 3px rgba(255, 255, 255, .6),
                    inset 2px 2px 3px rgba(0, 0, 0, .6);
                }
            }
        }

        ul {
            list-style-type: none;
            padding: 0;

            li {
                display: flex;
                flex-direction: row;
                justify-content: space-between;
                margin-bottom: 15px;
                font-size: 1.4rem;
                background-color: lightgrey;
                padding: 0.7rem;
                border-radius: 5px;
                overflow-wrap: anywhere;
                box-shadow: 10px 9px 12px -7px rgba(29,29,29,0.75);
                transition: transform .2s;

                &:hover {
                    transform: scale(1.03);

                    .customColorPicker {
                        display: block;
                    }
                }
            }
        }

        .gridView {
            display: flex;
            flex-wrap: wrap;

            li {
                margin: 0.4rem;
                min-width: 221px;
                max-width: 221px;
            }
        }

        .customListTools_container {
            display: flex;
            flex-direction: row;
            align-items: center;
            max-height: 28px;
            margin: auto 0.1rem auto 0.5rem;
            position: relative;

            .customColorPicker {
                display: none;
                cursor: pointer;   
                position: absolute;
                right: 35px;
            }

            .removeButton {
                background-color: lightcoral;
                color: white;
                border: none;
                font-weight: bold;
                border-radius: 50%;
                padding: 7px;
                cursor: pointer;
    
                &:focus, &:hover {
                    outline: none;
                    background-color: red;
                }
            }
        }
    }

    @media only screen and (max-width: 570px) {
        .viewAs_mbl {
            display: none;
        }
    }
}