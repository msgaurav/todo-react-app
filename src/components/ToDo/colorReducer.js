function colorReducer(state, action) {
  switch (action.type) {
  case 'fillListColor':
    return action.data
  case 'resetColor':
    return action.data
  default:
    throw new Error('unexpected action type')
  }
}

export default colorReducer
