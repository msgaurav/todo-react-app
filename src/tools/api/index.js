import React, { createContext } from 'react'
import PropTypes from 'prop-types'

// Create global contextAPI
export const UserContext = createContext()

// This context provider is passed to any component requiring the context
export function UserProvider({ children }) {
  // Context values
  const configValue = {
    showColorPicker: true,
  }

  return (
    <UserContext.Provider value={configValue}>
      {children}
    </UserContext.Provider>
  )
}

UserProvider.propTypes = {
  children: PropTypes.node.isRequired,
}
