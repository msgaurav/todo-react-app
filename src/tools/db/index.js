// App props data
const PropsData = {
  headerData: {
    heading    : 'ToDo App',
    description: 'Quick to add tasks, organize them into lists',
  },
}

export default PropsData
