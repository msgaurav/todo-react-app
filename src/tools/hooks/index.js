/* eslint-disable import/prefer-default-export */

// Single source of truth for custom hooks export
export { default as useWindowDimensions } from './useWindowDimensions'
