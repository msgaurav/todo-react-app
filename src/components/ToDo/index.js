import React, {
  useContext, useEffect, useReducer, useState,
} from 'react'
import ClearIcon from '@material-ui/icons/Clear'
import { useWindowDimensions } from '../../tools/hooks'
import { UserContext } from '../../tools/api'
import { Button } from '..'
import colorReducer from './colorReducer'
import styles from './ToDo.module.scss'

export default function ToDo() {
  const [inputValue, setInputValue] = useState('')
  const [toDoList, setToDoList] = useState([])
  const [uid, setUid] = useState(0)
  const [isView, setIsView] = useState(false)
  const [listColor, dispatchColor] = useReducer(colorReducer, null)

  // Destructuring useWindowDimensions hooks value
  const { windowWidth } = useWindowDimensions()

  // Destructuring useContext value
  const { showColorPicker } = useContext(UserContext)

  const handleInput = (e) => {
    e.preventDefault()
    setInputValue(e.target.value)
  }

  const handleSubmit = (e) => {
    e.preventDefault()
    setToDoList([
      ...toDoList,
      {
        id   : uid,
        input: inputValue,
      },
    ])
    setUid(uid + 1)
    setInputValue('')
  }

  const handleListView = () => {
    setIsView(!isView)
  }

  const handleBrush = (e) => {
    e.preventDefault()
    const brushColor = e.target.value
    const targetElement = e.target.parentElement.parentElement
    targetElement.style.backgroundColor = brushColor
  }

  const handleRemover = (e) => {
    e.preventDefault()
    // eslint-disable-next-line no-underscore-dangle
    const _uid = e.target.id
    setToDoList(toDoList.filter((element) => element.id !== Number(_uid)))
  }

  const handleResetBtn = () => {
    // eslint-disable-next-line no-alert
    const answer = window.confirm('This will clear your Todo List?')
    if (answer) {
      setUid(0)
      setToDoList([])
      setIsView(false)
      setInputValue('')
      dispatchColor({
        type: 'resetColor',
        data: null,
      })
    }
  }

  useEffect(() => {
    if (windowWidth < 570) {
      setIsView(false)
    }
  }, [windowWidth])

  return (
    <div className={styles.ToDo_wrapper}>
      <form onSubmit={handleSubmit} className={styles.form_container}>
        <input
          className={styles.inputForm_element}
          type="text"
          placeholder="Enter your task here"
          onChange={handleInput}
          value={inputValue}
          required
        />

        {inputValue.length > 0 && (
          <ClearIcon
            className={styles.clearInput_icon}
            onClick={() => setInputValue('')}
          />
        )}
      </form>

      {toDoList.length > 0 && (
        <div className={styles.taskList_container}>
          <div className={styles.tools_container}>
            {showColorPicker && (
              <input
                type="color"
                className={styles.colorPicker}
                title="Pick a Color"
                value={listColor === null ? '#4682b4' : listColor}
                onChange={(e) => dispatchColor({
                  type: 'fillListColor',
                  data: e.target.value,
                })}
              />
            )}

            <Button
              title="View as"
              className={styles.viewAs_mbl}
              onClick={handleListView}
            />

            <Button
              title="Reset"
              onClick={handleResetBtn}
            />
          </div>

          <ul className={isView ? styles.gridView : ''}>
            {toDoList.map((element, index) => (
              <li
                key={element.id}
                id={element.id}
                style={{ backgroundColor: listColor }}
              >
                <div>
                  <span>
                    {index + 1}
                    .&nbsp;
                  </span>
                  <span>{element.input}</span>
                </div>

                <div className={styles.customListTools_container}>
                  <input
                    id={element.id}
                    type="color"
                    title="Pick a Color"
                    className={styles.customColorPicker}
                    onChange={(e) => handleBrush(e)}
                  />

                  <button
                    type="button"
                    id={element.id}
                    className={styles.removeButton}
                    title="Delete"
                    onClick={handleRemover}
                  >
                    X
                  </button>
                </div>
              </li>
            ))}
          </ul>
        </div>
      )}
    </div>
  )
}
