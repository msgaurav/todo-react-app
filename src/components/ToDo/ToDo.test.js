import React from 'react'
import { render } from '@testing-library/react'
import { UserContext } from '../../tools/api'
import ToDo from './index'

const TestComponent = () => (
  <UserContext.Provider value>
    <ToDo />
  </UserContext.Provider>
)

describe('The ToDo Component', () => {
  it('snapshot renders correctly', () => {
    const { asFragment } = render(
      <TestComponent />,
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('should renders with correct placeholder', () => {
    const { getByPlaceholderText } = render(
      <TestComponent />,
    )
    expect(getByPlaceholderText('Enter your task here')).toBeInTheDocument()
  })
})
