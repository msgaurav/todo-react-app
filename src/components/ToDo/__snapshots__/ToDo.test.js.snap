// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ToDo Component snapshot renders correctly 1`] = `
<DocumentFragment>
  <div
    class="ToDo_wrapper"
  >
    <form
      class="form_container"
    >
      <input
        class="inputForm_element"
        placeholder="Enter your task here"
        required=""
        type="text"
        value=""
      />
    </form>
  </div>
</DocumentFragment>
`;
